### ISSUE TITLE
KNIME Analytics Platform prompts for proxy user and password on startup

### ISSUE THREAD

**Issue by kickuth** [knime-mpicbg/knime-scripting#109]
When starting the Analytics Platform with a proxy requiring authentication configured, the user will be prompted to enter proxy credentials:

![image](https://user-images.githubusercontent.com/12065930/162446355-6e8f2417-ce70-464c-961a-ce070709600a.png)

(verify with e.g. [Docker Hub](https://hub.docker.com/r/robhaswell/squid-authenticated): docker run -e SQUID_USERNAME=testuser -e SQUID_PASSWORD=pwtest -p 3128:3128 robhaswell/squid-authenticated)

This seems to be caused by file downloads from [HERE](https://github.com/knime-mpicbg/knime-scripting/blob/ba785d914825c72f5b849ff5cbd5d3c52dbc9d0b/de.mpicbg.knime.scripting.python/src/de/mpicbg/knime/scripting/python/prefs/PythonPreferenceInitializer.java#L66). While the prompt is open, a call stack reveals what we are waiting on:

```
at de.mpicbg.knime.scripting.core.TemplateCache.pingURL(TemplateCache.java:404)
at de.mpicbg.knime.scripting.core.TemplateCache.isFileReachable(TemplateCache.java:382)
at de.mpicbg.knime.scripting.core.TemplateCache.addTemplateFile(TemplateCache.java:126)
at de.mpicbg.knime.scripting.core.TemplateCache.addTemplatesFromPref(TemplateCache.java:105)
at de.mpicbg.knime.scripting.core.utils.ScriptingUtils.loadTemplateCache(ScriptingUtils.java:39)
```

depending on the installed plugin either:
```
at de.mpicbg.knime.scripting.python.PythonScriptingBundleActivator.start(PythonScriptingBundleActivator.java:67)
at de.mpicbg.knime.scripting.r.R4KnimeBundleActivator.start(R4KnimeBundleActivator.java:68)
```

I hope it is possible to avoid the downloads or only trigger them once the proxy user/password information can be retrieved from e.g. the eclipse store.

**Comment by kickuth** [comment]
Just noticed that these downloads are likely also made by the other scripting extensions.

Any idea, how to prevent this? Not all proxies return information (see screenshot), so that inexperienced AP users in companies get this (to them) cryptic startup prompt and can't continue.

![image](https://user-images.githubusercontent.com/12065930/164195369-0f8765aa-d95e-412b-a7c9-992cd11bbac6.png)

**Comment by niederle** [comment]
Thank you for reporting the issue. I'll give it a look as soon as possible.

**Comment by bernd-wiswedel** [comment]
I need to bump this up again. There are users having proxy configurations and their KNIME installation would not start up because of this.

I think the proper solution is to defer [this line](https://github.com/knime-mpicbg/knime-scripting/blob/develop/de.mpicbg.knime.scripting.python/src/de/mpicbg/knime/scripting/python/PythonScriptingBundleActivator.java#L67) in `PythonScriptingBundleActivator` to the latest time possible (i.e. when the nodes use/need the templates, i.e. pull from the cache lazily).

**Comment by niederle** [comment]
@bernd-wiswedel 
I wanted to put it into the first configuration call of one of the nodes. But I realized when working with the SDK KNIME loads a previously open workflow which also means that the configuration of 'configured' nodes is called during startup. Would this be an issue or does the Welcome-page always prevent it? I am not sure about this.

**Comment by bernd-wiswedel** [comment]
I think that would be acceptable given that the user is using nodes from the scripting extension. Also, as you mentioned, the workflows only get loaded when the welcome page is closed (this might change in the future but that's a discussion for another time).

On a related note, I think there is currently no timeout when fetching the scripts -- that's something to change. 

You can reproduce this by setting up a local proxy container (see @kickuth 's comment above) and configure KNIME's proxy configuration (File -> Preferences -> General -> Network Settings) to use that proxy by provide the wrong password. Upon restart, KNIME will freeze indefinitely.

**Comment by niederle** [comment]
@bernd-wiswedel 

> On a related note, I think there is currently no timeout when fetching the scripts -- that's something to change.

There should be a timeout. Without any network connection KNIME will start without problems. I'm pretty sure that this has been tested from my side.

> You can reproduce this by setting up a local proxy container (see @kickuth 's comment above) and configure KNIME's proxy configuration (File -> Preferences -> General -> Network Settings) to use that proxy by provide the wrong password. Upon restart, KNIME will freeze indefinitely.

I cannot reproduce the freezing. When setting a proxy with a wrong password, I get another dialog which prompts for username and password. Then providing the correct credentials would solve the issue. I am working on a Mac. I have not tested it on other OS. On which operating system did this issue occur?

**Comment by bernd-wiswedel** [comment]
I'm on Linux. I can't currently reproduce either ( :-( ). 

(Happy to validate your (other) changes once they are made.)

**Comment by niederle** [comment]
@bernd-wiswedel 

Feel free to test my solution on our trunk build. I'm happy about feedback.

**Comment by bernd-wiswedel** [comment]
I can confirm the behavior is as expected now. KNIME starts without issues and a connection is (attempted to be) established only when I interact with the nodes.

I noticed that the extension is not signed (which I guess is known for trunk/master builds). I followed the instructions [here](https://www.knime.com/community), subsection "Nightly Builds".

Also, I assume you will have to take additional steps in order to make this fix available in the next iteration of KNIME AP (4.7). Tagging @gab1one here so that he's aware of it. (Release branch must not be copied but a new one needs to created off of master?)

Thank you.
